### PR TITLE
Add an interface to mark the solution as invalid / out-of-bounds

### PR DESCRIPTION
**Pull request by MengnanLi91**
ref #22814 Adding new feature to test whether the material property is valid

## Reason

The feature is requested to determine whether the solution is valid at the time of convergence.

## Design

The function can be called from materials to state the converged solution is invalid. 
A bool function is added with a flag. The default value of the flag is false meaning solution is valid. The flag will flip every time the function is called. The final value will be the state of the converged solution.

## Impact

The new feature can state the converged solution is invalid.

### CODE REVIEW

**Comment by friedmud** [comment]
I should mention that this is not the _final_ solution for this issue... but it is _a_ solution that will close the ticket.

**Review by GiudGiud — CHANGES_REQUESTED** [review]
PR looks pretty good!
Some suggestions

While it s part of the solution to tag the solution as invalid, it s only on the final residual evaluation that we should do that. The Newton solve often visits numerous invalid states before converging on a valid one

So the 'invalid' status needs to be reset right before this final evaluation.

**Comment by GiudGiud** [comment]
You ll want to rebase on upstream/devel instead of merging btw

**Comment by friedmud** [comment]
@GiudGiud the flag resets before every residual evaluation... so the state of the flag at the end of the solve will represent what happened with the final solution.

**Comment by GiudGiud** [comment]
Umm I must have missed that. Can one of you please point me to the line which does that?

**Comment by friedmud** [comment]
@GiudGiud - hah!  You're right... the reset didn't make it into the PR.  It existed... but didn't make it into the commits!

File: framework/src/systems/NonlinearSystem.C
Comment (GiudGiud) [reply framework/src/systems/NonlinearSystem.C:346]: do you think we could cover this with just a set of names with an insert

the set would be reset at the same time as the boolean for invalid

File: framework/include/interfaces/SolutionInvalidInterface.h
Comment (MengnanLi91) [reply framework/include/interfaces/SolutionInvalidInterface.h:21]: This is updated.

File: framework/include/interfaces/SolutionInvalidInterface.h
Comment (MengnanLi91) [reply framework/include/interfaces/SolutionInvalidInterface.h:35]: This is updated.

File: framework/include/systems/SystemBase.h
Comment (MengnanLi91) [reply framework/include/systems/SystemBase.h:910]: This is updated.

File: framework/include/systems/SystemBase.h
Comment (MengnanLi91) [reply framework/include/systems/SystemBase.h:911]: This function is often used with if condition(e.g. if(solutionInvalid) do something;), I feel statement without "is" is less confusing. What do you think?

File: framework/include/systems/SystemBase.h
Comment (MengnanLi91) [reply framework/include/systems/SystemBase.h:913]: Good point. This is updated.

File: framework/src/systems/NonlinearSystem.C
Comment (MengnanLi91) [reply framework/src/systems/NonlinearSystem.C:346]: I agree. This is the part I am working on. It will be added in later version of this feature.

File: test/include/materials/NonsafeMaterial.h
Comment (MengnanLi91) [reply test/include/materials/NonsafeMaterial.h:16]: This is updated

File: test/tests/materials/solution_invalid/tests
Comment (MengnanLi91) [reply test/tests/materials/solution_invalid/tests:8]: This is updated.

File: framework/include/systems/SystemBase.h
Comment (GiudGiud) [reply framework/include/systems/SystemBase.h:910]: seems fine now with the other renamed

File: framework/include/interfaces/SolutionInvalidInterface.h
Comment (GiudGiud) [inline framework/include/interfaces/SolutionInvalidInterface.h:21]:
```suggestion
 * An interface to communicate a solution invalid state to FEProblemBase
```

File: framework/include/interfaces/SolutionInvalidInterface.h
Comment (GiudGiud) [inline framework/include/interfaces/SolutionInvalidInterface.h:33]:
```suggestion
  /// A pointer to the FEProblem
```

**Comment by GiudGiud** [comment]
Another thing that will help a lot to deploy this everywhere is an option to bypass it, either on a Problem scale or for each material. 
‘basically a dont_set_invalid parameter but with a better name.

**Comment by MengnanLi91** [comment]
> 

Good point! I'll add it to the PR I am working on along with object tracking.

**Comment by friedmud** [comment]
@GiudGiud - I think this is good to go.  We are already working on a follow-up PR where we are adding a lot of features.  Let's get this one in and close the "safety" ticket so @cticenhour can breathe easily!

**Review by GiudGiud — DISMISSED** [review]
still no reset mechanism.

This is not usable as is. No app is going to want to invalidate their solution with no recourse (the boolean I asked for) and no information about what happened? (the list of objects involved, reminder that the warnings are being silenced)

Also the test is crashing on the warning, not on the return false on accepting the solution. It's not just missing a line, it's not testing what we want.

**Comment by GiudGiud** [comment]
@cticenhour please review for concurrence

**Comment by cticenhour** [comment]
OK @GiudGiud - will chime in this morning in between meetings.

**Comment by friedmud** [comment]
Bah - I thought it had been fixed up.  My bad.

**Review by cticenhour — DISMISSED** [review]
I think this is the bare minimum required to satisfy the original issue now that the reset mechanism was just put in place. I have a few suggestions to make code comments and other quality-of-life stuff better. 

I agree with @GiudGiud that some "opt out" scenario should be available to those who want to implement this in their object and not use it for special cases - this is a vital piece of the feature set to me for it to be useful for apps. I disagree that it *needs* to be in this PR, but that disagreement is contingent on the ETA for the follow-up. When is this expected, and it is already being worked on in another branch?

**Comment by cticenhour** [comment]
Note for all on the current set of test results so far - package builds will be failing until we get a next --> devel merge.

**Comment by milljm** [comment]
You can also rebase on Next if you wish to not wait...

File: large_media
Comment (cticenhour) [reply large_media:1]: This regression on the `large_media` submodule status needs to be removed from the commit history. Seems it was accidentally added.

Comment (cticenhour) [reply large_media]: Resolved in d549645

File: framework/include/interfaces/SolutionInvalidInterface.h
Comment (loganharbour) [inline framework/include/interfaces/SolutionInvalidInterface.h:34]: This should be a reference, not a pointer

**Comment by friedmud** [comment]
@GiudGiud @cticenhour 

The branch is all rebased and squashed.  It has doco for the new interface and adds to the newsletter.  We added the ability to turn this check off.

As for why there isn't more information coming out: it's because it's not as straightforward as you think.  You cannot simply pass a string in to print out or add to a std::set or whatever - that is way too slow.  Think of the case where you needed to call this function on thousands (or millions) of quadrature points... it would _crawl_ if you were doing string operations.  What's needed is to have a registration system (similar to PerfGraph sections) so that you are just keeping track of counts of unique IDs.

That's what @MengnanLi91 is implementing now.  She's already a long way down the path due to picking up a generalized registration that @loganharbour had made a while ago.  The new system will (efficiently) keep track of how many times each piece of code set the solution to be invalid... and _when_... and be able to print it all out.

She's already a long way down the path - so I'm hopeful we will have it in by the end of next week.

THAT is why I don't want to put a ton of work into this PR - what is coming is WAY better... but this is good enough for now.

d08b19c28aa886d6f7cf314fa32532b7

**Comment by cticenhour** [comment]
@friedmud I see your point - and really like the idea you are broadly proposing compared to this one. I am willing to approve this pending @loganharbour's comment being addressed (`_si_fe_problem` really should be a reference, not a pointer, but I missed that earlier).

**Review by GiudGiud — COMMENTED** [review]
same approval as Casey

if you assume it would be done poorly, then yes it would be slow.
It s easy to imagine a workable solution that is not slow: keep track of the status with a boolean on each material, then do the "string" operation only when outputting the error message.

**Comment by MengnanLi91** [comment]
> 

LazyCoupleable.h use pointer while MortarConsumerInterface.h use reference for FEProblemBase. I am confused now...

**Comment by loganharbour** [comment]
> > 
> 
> @cticenhour @loganharbour LazyCoupleable.h use pointer while MortarConsumerInterface.h use reference for FEProblemBase. I am confused now...

`LazyCouplable` was developed... lazily :-)

In general - you should only need to use a pointer if there is a valid reason for it to be null. In this case, we get an address to the problem on construction and it's always available. Therefore, a reference is more appropriate. Only use a pointer if a reference won't work for the situation.

This'll happen in time unfortunately - that is - finding old code where poor development choices were made. We don't have the resources to rewrite the code base with modern standards, but we do have the time to make sure that _new_ code is up to higher standards.

**Comment by MengnanLi91** [comment]
> > > 
> > 
> > 
> > @cticenhour @loganharbour LazyCoupleable.h use pointer while MortarConsumerInterface.h use reference for FEProblemBase. I am confused now...
> 
> `LazyCouplable` was developed... lazily :-)
> 
> In general - you should only need to use a pointer if there is a valid reason for it to be null. In this case, we get an address to the problem on construction and it's always available. Therefore, a reference is more appropriate. Only use a pointer if a reference won't work for the situation.
> 
> This'll happen in time unfortunately - that is - finding old code where poor development choices were made. We don't have the resources to rewrite the code base with higher standards, but we do have the time to make sure that _new_ code is up to higher standards.

Got it. I'll change it to a reference

**Comment by cticenhour** [comment]
@loganharbour beat me to it while I was still typing, so I will delete and chime in with only this :) 

`LazyCoupleable` was expected not to have an address for `FEProblem` until after it was constructed, so a pointer was used so that it could be null until it was set.

**Comment by friedmud** [comment]
@loganharbour @cticenhour I disagree about the pointer business BTW.  The actual rule in MOOSE is that you try to avoid changing a pointer to a reference or a reference to a pointer.  If you receive something as a pointer then store/return it as one.  If you receive something as a reference then store/return it as one.

In this case the pointer is being received from the parameters as a pointer and there is nothing wrong with storing it as a pointer... especially since it's a private member of a rather small class.

We really need to loosen up on this kind of thing.  Let's debate stuff that matters, like public interfaces that users will see...

@GiudGiud , this ultimately won't be just for Materials.  An object might (probably) will have multiple different correlations that could be out of bounds.  You don't only want to know which object - you also want to know which correlation and get a reasonable print out about that particular correlation and what could be wrong.  Doing that requires a whole system for registering each place where `setSolutionInvalid()` is called with a good message to go along with it.

**Comment by loganharbour** [comment]
> @loganharbour @cticenhour I disagree about the pointer business BTW. The actual rule in MOOSE is that you try to avoid changing a pointer to a reference or a reference to a pointer. 

This doesn’t make a shred of sense. We even check if the pointer is valid… there’s no reason to use a pointer in the slightest

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Thank you for the patch @MengnanLi91 !

Congratulations on your first PR merged to MOOSE!

**Comment by friedmud** [comment]
@GiudGiud - thanks for the merge!

File: framework/include/problems/FEProblemBase.h
Comment (lindsayad) [inline framework/include/problems/FEProblemBase.h:1620]: Can be `const`

Comment (MengnanLi91) [reply framework/include/problems/FEProblemBase.h:1620]: I'll update this in the next version of the feature.

Comment (lindsayad) [reply framework/include/problems/FEProblemBase.h:1620]: I'm very impressed with our quickly you're contributing @MengnanLi91!

Comment (lindsayad) [reply framework/include/problems/FEProblemBase.h:1620]: It must be the NCSU power 😄

**Comment by lindsayad** [comment]
I do think that in general it communicates a lot more safety to a developer when they are dealing with a reference instead of a pointer. When I get a pointer from an API my first inkling it to `mooseAssert(pointer, "This better not be null as I'm about to use it!");`

File: framework/include/problems/FEProblemBase.h
Comment (MengnanLi91) [reply framework/include/problems/FEProblemBase.h:1620]: Go wolfpack! I also got a lot of help from our awesome team to get my first PR done : )
image

**Comment by MengnanLi91** [comment]
@GiudGiud I am working on improving the feature as we discussed. Below is the output looks like. Does it meet your need? Or is there any other features I should add?
Screenshot 2022-12-15 at 3 08 18 PM
Screenshot 2022-12-15 at 5 07 43 PM

**Comment by GiudGiud** [comment]
We should get feedback from the analysis groups on this at some point

I think we would want to show messages instead of sections? 
a computeQpProperties could trigger solutionInvalid in multiple locations

**Comment by MengnanLi91** [comment]
> We should get feedback from the analysis groups on this at some point
> 
> I think we would want to show messages instead of sections? a computeQpProperties could trigger solutionInvalid in multiple locations

Below are the current version I have. What do you think? 
1. When there is solution invalid warning detected but allow_invalid_solution is true, the output is
image
2. When there is solution invalid warning detected and not allowed, the output is
image
3. When there is no solution invalid warning detected or no section registered for solution invalid check, the output is
image

**Comment by GiudGiud** [comment]
Looks good.
So the 'Total iterations' it's the number of occurences? Over all linear, nonlinear and timesteps?
Or the number of iterations that presented the issue

**Comment by MengnanLi91** [comment]
> Looks good.
> 
> So the 'Total iterations' it's the number of occurences? Over all linear, nonlinear and timesteps?
> 
> Or the number of iterations that presented the issue

The total iterations count all warnings over all linear, nonlinear iterations and time steps while the last iteration will be reset for each iteration. Is it what we want to have? If this is okay,I could push this version for code review.

**Comment by GiudGiud** [comment]
This total is somewhat obscure for users though. Users dont know that we take a residual evaluation step at the beginning and end of every time step for example.
It's a sum of residual and jacobian loops too, and that mapping to linear and non linear is somewhat obscure to users. 

I would say only give the final iteration # of occurrences

**Pull request merged**
